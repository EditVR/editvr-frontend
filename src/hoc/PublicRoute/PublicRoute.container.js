/**
 * @file PublicRoute.container.js
 * Exports a redux-connected PublicRoute component.
 */

import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';

import PublicRoute from './PublicRoute';

const mapStateToProps = state => ({
  authentication: state.user.authentication
});

export default connect(mapStateToProps)(withRouter(PublicRoute));
