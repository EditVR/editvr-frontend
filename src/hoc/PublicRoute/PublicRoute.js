/**
 * @file PublicRoute.js
 * Exports a route handler that makes a given page exclusively public.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Route, Redirect } from 'react-router-dom';

const PublicRoute = ({
  component: Component,
  authentication,
  location,
  redirectTo,
  ...rest
}) => (
  <Route
    {...rest}
    render={props => {
      const { accessToken, created, expiresIn } = authentication;

      // If the access token does not exist, or the token has expired, false.
      let isAuthenticated = true;
      if (!accessToken || Date.now() - created > expiresIn * 1000) {
        isAuthenticated = false;
      }

      return isAuthenticated ? (
        <Redirect push to={redirectTo} />
      ) : (
        <Component {...props} />
      );
    }}
  />
);

PublicRoute.propTypes = {
  component: PropTypes.func.isRequired,
  redirectTo: PropTypes.string.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string.isRequired,
    search: PropTypes.string,
    hash: PropTypes.string
  }).isRequired,
  authentication: PropTypes.shape({
    accessToken: PropTypes.string,
    created: PropTypes.number,
    expiresIn: PropTypes.number
  })
};

PublicRoute.defaultProps = {
  authentication: {}
};

export default PublicRoute;
