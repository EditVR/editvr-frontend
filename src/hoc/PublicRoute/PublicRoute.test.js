/**
 * @file PublicRoute.test.js
 * Contains tests for PublicRoute.js.
 */

import React, { Fragment } from 'react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';

import PublicRoute from './PublicRoute.container';

describe('<PublicRoute />', () => {
  it('If a user is not authenticated, PublicRoute will render its specified component.', () => {
    const store = configureStore()({
      user: {
        authentication: {}
      }
    });

    expect(
      renderer
        .create(
          <Provider store={store}>
            <Router initialEntries={['/dashboard']}>
              <Fragment>
                <PublicRoute
                  path="/dashboard"
                  component={() => <div>test component</div>}
                />
              </Fragment>
            </Router>
          </Provider>
        )
        .toJSON()
    ).toMatchSnapshot();
  });

  it('If a user is authenticated, PublicRoute will redirect to the user dashboard.', () => {
    const store = configureStore()({
      user: {
        authentication: {
          accessToken: 'stub',
          refreshToken: 'stub',
          csrfToken: 'stub',
          expiresIn: 30000,
          created: Date.now()
        }
      }
    });

    expect(
      renderer
        .create(
          <Provider store={store}>
            <Router initialEntries={['/dashboard']}>
              <Fragment>
                <PublicRoute
                  path="/dashboard"
                  component={() => <div>test component</div>}
                />
              </Fragment>
            </Router>
          </Provider>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
