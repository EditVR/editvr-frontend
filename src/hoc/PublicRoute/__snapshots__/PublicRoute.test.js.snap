// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PublicRoute /> If a user is authenticated, PublicRoute will redirect to the specified redirect location. 1`] = `null`;

exports[`<PublicRoute /> If a user is not authenticated, PublicRoute will render its specified component. 1`] = `
<div>
  test component
</div>
`;
